### PR TITLE
Review edit distance

### PR DESCRIPTION
**Pull request by marehr**
@h-2 as promised I addressed all points in your code review from #431

Can you have a second look?

Can you squash merge this if done?

### CODE REVIEW

File: include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp [hunk 1]
```diff
@@ -334,59 +331,68 @@ struct alignment_matrix_formatter
...
     //!\brief Convert a matrix entry into a std::string
-    std::string as_string(auto && entry) const noexcept
+    static std::string as_string(auto && entry) noexcept
     {
         std::stringstream stream;
-        if constexpr (alphabet_concept<remove_cvref_t<decltype(entry)>> &&
-                      !ostream_concept<std::stringstream, decltype(entry)>)
-        {
+        if constexpr (alphabet_concept<remove_cvref_t<decltype(entry)>>)
             stream << to_char(entry);
```
Comment (marehr) [inline include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp]: @h-2 Can I use the debugstream here, which overloads stringstream?

Comment (h-2) [reply include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp]: yes

**Comment by h-2** [comment]
Cool, I hope to find some time this week for it!

**Comment by h-2** [comment]
Thanks for looking into all the small comments, but one major question remains: can we get rid of the matrix_formatter and instead just overload the debug_stream for row_wise_matrix? The debug stream has a UTF8 knob know so you can do UTF8-based pretty printing. But it does mean you have to decide on *one style* for pretty printing. However I think having non-UTF8 and UTF8 as choices is already pretty good considering that this is only debug functionality!

**Comment by marehr** [comment]
> Thanks for looking into all the small comments, but one major question remains: can we get rid of the matrix_formatter and instead just overload the debug_stream for row_wise_matrix? The debug stream has a UTF8 knob know so you can do UTF8-based pretty printing. But it does mean you have to decide on _one style_ for pretty printing. However I think having non-UTF8 and UTF8 as choices is already pretty good considering that this is only debug functionality!

Yeah sure, I can workaround it. But, can we do this in a new PR? I already opened an issue #510 for this that I will address soon. Maybe you can explain me how to access this `UTF8 knob`.

**Comment by h-2** [comment]
> Yeah sure, I can workaround it. But, can we do this in a new PR? I already opened an issue #510 for this that I will address soon. 

Ok!

> Maybe you can explain me how to access this UTF8 knob.

See http://docs.seqan.de/seqan/3.0.0-master-dev/classseqan3_1_1debug__stream__type.html and http://docs.seqan.de/seqan/3.0.0-master-dev/group__stream.html#ga3f0d33f7d8148aff69c8dbee851695a2

So when printing, you can just do: 
```cpp
debug_stream << fmtflags2::utf8 << your_matrix;
```
and it will print with the UTF8 symbols.

**Comment by marehr** [comment]
Ah nice. This is handy 😀

Am 16. Oktober 2018 13:00:15 MESZ schrieb Hannes Hauswedell <notifications@github.com>:
>> Yeah sure, I can workaround it. But, can we do this in a new PR? I
>already opened an issue #510 for this that I will address soon. 
>
>Ok!
>
>> Maybe you can explain me how to access this UTF8 knob.
>
>See
>http://docs.seqan.de/seqan/3.0.0-master-dev/classseqan3_1_1debug__stream__type.html
>and
>http://docs.seqan.de/seqan/3.0.0-master-dev/group__stream.html#ga3f0d33f7d8148aff69c8dbee851695a2
>
>So when printing, you can just do: 
>```cpp
>debug_stream << fmtflags2::utf8 << your_matrix;
>```
>and it will print with the UTF8 symbols.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/seqan/seqan3/pull/495#issuecomment-430194467

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

File: include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp [hunk 2]
```diff
@@ -334,59 +332,66 @@ struct alignment_matrix_formatter
             if (row == 0)
                 print_first_cell(symbols.epsilon);
             else
-                print_first_cell(query[row-1]);
-            for(unsigned col = 0; col < matrix.cols(); ++col)
+                print_first_cell(as_string(query[row - 1]));
+            for (size_t col = 0; col < matrix.cols(); ++col)
                 print_cell(entry_at(row, col));
             cout << "\n";
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp]: This is a local stream that is not `std::cout`.

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp [hunk 3]
```diff
@@ -329,22 +329,22 @@ struct pairwise_alignment_edit_distance_unbanded
...
-    inline bool prev_last_active_cell()
+    bool prev_last_active_cell()
     {
         score_mask >>= 1;
         if (score_mask != (word_type)0)
             return true;
 
         last_block--;
-        if (is_global && last_block == (unsigned)-1)
+        if (is_global && last_block == (size_t) - 1)
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: I want to detect an underflow. `(size_t) - 1` yields me the largest number.

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: I changed the check

File: include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/alignment/matrix/alignment_matrix_formatter.hpp]: but why call it like something in the standard library? Just to confuse me? o_O

**Pull request merged**